File: src/pwa/composables/useComments.js
import { ref } from 'vue'
import { useApiStore } from '@/stores/api'
import { useAuthStore } from '@/stores/auth'

export function useComments() {
  const apiStore = useApiStore()
  const authStore = useAuthStore()
  
  const comments = ref([])
  const loading = ref(false)
  const error = ref(null)
  const currentType = ref(null)
  const currentId = ref(null)

  /**
   * Charger les commentaires pour un contenu donné
   */
  const loadComments = async (type = null, id = null) => {
    loading.value = true
    error.value = null
    
    // Si des paramètres sont fournis, les utiliser
    if (type && id) {
      currentType.value = type
      currentId.value = id
    }
    
    // Vérifier qu'on a les paramètres nécessaires
    if (!currentType.value || !currentId.value) {
      console.error('❌ [Comments] Type ou ID manquant pour charger les commentaires')
      error.value = 'Paramètres manquants pour charger les commentaires'
      loading.value = false
      return
    }
    
    try {
      console.log(`💬 [Comments] Chargement des commentaires pour ${currentType.value}:${currentId.value}`)
      
      const data = await apiStore.get('/comments', { 
        commentable_type: `App\\Models\\${currentType.value}`,
        commentable_id: currentId.value 
      })
      
      if (data.success) {
        comments.value = data.data || []
        console.log(`✅ [Comments] ${comments.value.length} commentaires chargés`)
      } else {
        comments.value = []
      }
    } catch (err) {
      console.error('❌ [Comments] Erreur lors du chargement:', err)
      error.value = err.message
      comments.value = []
    } finally {
      loading.value = false
    }
  }

  /**
   * Charger les commentaires avec des données fraîches (sans cache)
   */
  const loadCommentsFresh = async (type = null, id = null) => {
    loading.value = true
    error.value = null
    
    // Si des paramètres sont fournis, les utiliser
    if (type && id) {
      currentType.value = type
      currentId.value = id
    }
    
    // Vérifier qu'on a les paramètres nécessaires
    if (!currentType.value || !currentId.value) {
      console.error('❌ [Comments] Type ou ID manquant pour charger les commentaires')
      error.value = 'Paramètres manquants pour charger les commentaires'
      loading.value = false
      return
    }
    
    try {
      console.log(`💬 [Comments] Chargement frais des commentaires pour ${currentType.value}:${currentId.value}`)
      
      const data = await apiStore.getFresh('/comments', { 
        commentable_type: `App\\Models\\${currentType.value}`,
        commentable_id: currentId.value 
      })
      
      if (data.success) {
        comments.value = data.data || []
        console.log(`✅ [Comments] ${comments.value.length} commentaires chargés (frais)`)
      } else {
        comments.value = []
      }
    } catch (err) {
      console.error('❌ [Comments] Erreur lors du chargement frais:', err)
      error.value = err.message
      comments.value = []
    } finally {
      loading.value = false
    }
  }

  /**
   * Définir le contexte pour les commentaires
   */
  const setContext = (type, id) => {
    currentType.value = type
    currentId.value = id
    console.log(`💬 [Comments] Contexte défini: ${type}:${id}`)
  }

  /**
   * Ajouter un nouveau commentaire
   */
  const addComment = async (type, id, content, authorName = null, authorEmail = null) => {
    if (!content?.trim()) {
      throw new Error('Le contenu du commentaire ne peut pas être vide')
    }

    // Permettre les commentaires anonymes avec nom et email
    if (!authStore.isAuthenticated && (!authorName || !authorEmail)) {
      throw new Error('Veuillez fournir votre nom et email pour commenter en mode anonyme')
    }

    try {
      console.log(`💬 [Comments] Ajout d'un commentaire pour ${type}:${id}`)
      
      const commentData = {
        commentable_type: `App\\Models\\${type}`,
        commentable_id: parseInt(id),
        content: content.trim()
      }

      // Ajouter les données anonymes si nécessaire
      if (!authStore.isAuthenticated) {
        commentData.author_name = authorName
        commentData.author_email = authorEmail
      }

      const data = await apiStore.post('/comments', commentData)
      
      if (data.success) {
        console.log('✅ [Comments] Commentaire ajouté avec succès')
        
        // Ajouter le nouveau commentaire à la liste locale
        const newComment = {
          id: data.data?.id || Date.now(),
          content: content.trim(),
          author_name: authStore.userName || 'Utilisateur',
          created_at: new Date().toISOString(),
          ...data.data
        }
        
        comments.value.unshift(newComment)
        return newComment
      } else {
        throw new Error(data.message || 'Erreur lors de l\'ajout du commentaire')
      }
    } catch (err) {
      console.error('❌ [Comments] Erreur lors de l\'ajout:', err)
      
      // Si erreur d'authentification, informer l'utilisateur
      if (err.message.includes('401') || err.message.includes('Unauthorized')) {
        throw new Error('Votre session a expiré. Veuillez vous reconnecter.')
      }
      
      throw err
    }
  }

  /**
   * Supprimer un commentaire (si autorisé)
   */
  const deleteComment = async (commentId) => {
    if (!authStore.isAuthenticated) {
      throw new Error('Vous devez être connecté pour supprimer un commentaire')
    }

    try {
      console.log(`💬 [Comments] Suppression du commentaire ${commentId}`)
      
      const data = await apiStore.del(`/comments/${commentId}`)
      
      if (data.success) {
        console.log('✅ [Comments] Commentaire supprimé avec succès')
        
        // Retirer le commentaire de la liste locale
        const index = comments.value.findIndex(c => c.id === commentId)
        if (index > -1) {
          comments.value.splice(index, 1)
        }
        
        return true
      } else {
        throw new Error(data.message || 'Erreur lors de la suppression du commentaire')
      }
    } catch (err) {
      console.error('❌ [Comments] Erreur lors de la suppression:', err)
      throw err
    }
  }

  /**
   * Formater la date d'un commentaire
   */
  const formatCommentDate = (date) => {
    if (!date) return ''
    
    try {
      const commentDate = new Date(date)
      const now = new Date()
      const diffInMinutes = Math.floor((now - commentDate) / (1000 * 60))
      
      if (diffInMinutes < 1) return 'À l\'instant'
      if (diffInMinutes < 60) return `Il y a ${diffInMinutes}mn`
      
      const diffInHours = Math.floor(diffInMinutes / 60)
      if (diffInHours < 24) return `Il y a ${diffInHours}h`
      
      const diffInDays = Math.floor(diffInHours / 24)
      if (diffInDays < 7) return `Il y a ${diffInDays}j`
      
      // Pour les dates plus anciennes, afficher la date complète
      return commentDate.toLocaleDateString('fr-FR', {
        day: 'numeric',
        month: 'short',
        year: commentDate.getFullYear() !== now.getFullYear() ? 'numeric' : undefined
      })
    } catch {
      return 'Date invalide'
    }
  }

  /**
   * Vérifier si l'utilisateur peut modifier/supprimer un commentaire
   */
  const canEditComment = (comment) => {
    if (!authStore.isAuthenticated || !comment) return false
    
    // L'utilisateur peut modifier son propre commentaire
    // ou si c'est un admin (à implémenter selon vos besoins)
    return comment.user_id === authStore.user?.id || authStore.user?.role === 'admin'
  }

  /**
   * Vérifier si l'utilisateur peut supprimer un commentaire
   */
  const canDeleteComment = (comment) => {
    if (!authStore.isAuthenticated || !comment) return false
    
    // L'utilisateur peut supprimer son propre commentaire
    // ou si c'est un admin (à implémenter selon vos besoins)
    return comment.user_id === authStore.user?.id || authStore.user?.role === 'admin'
  }

  return {
    // État
    comments,
    loading,
    error,
    
    // Actions
    loadComments,
    loadCommentsFresh,
    addComment,
    deleteComment,
    
    // Utilitaires
    formatCommentDate,
    canEditComment,
    canDeleteComment,
    setContext
  }
} 

File: flutter_app/lib/composables/use_comments.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../services/api_service.dart';

class CommentsNotifier extends StateNotifier<List<dynamic>> {
  final ApiService _apiService;
  
  CommentsNotifier(this._apiService) : super([]);

  Future<List<dynamic>> loadComments(String contextType, String contextId) async {
    try {
      print('📝 [CommentsNotifier] Chargement commentaires pour $contextType:$contextId');
      
      final response = await _apiService.get('/comments', params: {
        'commentable_type': contextType,
        'commentable_id': contextId,
      });
      
      if (response['success']) {
        final comments = response['data'] as List<dynamic>;
        state = comments;
        print('✅ [CommentsNotifier] ${comments.length} commentaires chargés');
        return comments;
      }
      
      return [];
    } catch (error) {
      print('❌ [CommentsNotifier] Erreur chargement commentaires: $error');
      return [];
    }
  }

  Future<List<dynamic>> loadCommentsFresh(String contextType, String contextId) async {
    try {
      print('🔄 [CommentsNotifier] Rechargement commentaires pour $contextType:$contextId');
      
      final response = await _apiService.request('/comments', forceRefresh: true, params: {
        'commentable_type': contextType,
        'commentable_id': contextId,
      });
      
      if (response['success']) {
        final comments = response['data'] as List<dynamic>;
        state = comments;
        print('✅ [CommentsNotifier] ${comments.length} commentaires rechargés');
        return comments;
      }
      
      return [];
    } catch (error) {
      print('❌ [CommentsNotifier] Erreur rechargement commentaires: $error');
      return [];
    }
  }

  Future<bool> addComment(String contextType, String contextId, String content) async {
    try {
      print('📝 [CommentsNotifier] Ajout commentaire pour $contextType:$contextId');
      
      final response = await _apiService.post('/comments', {
        'commentable_type': contextType,
        'commentable_id': contextId,
        'content': content,
      });
      
      if (response['success']) {
        // Recharger les commentaires après ajout
        await loadComments(contextType, contextId);
        print('✅ [CommentsNotifier] Commentaire ajouté avec succès');
        return true;
      }
      
      return false;
    } catch (error) {
      print('❌ [CommentsNotifier] Erreur ajout commentaire: $error');
      return false;
    }
  }

  Future<bool> deleteComment(String commentId) async {
    try {
      print('🗑️ [CommentsNotifier] Suppression commentaire ID: $commentId');
      
      final response = await _apiService.delete('/comments/$commentId');
      
      if (response['success']) {
        // Retirer le commentaire de la liste
        state = state.where((comment) => comment['id'].toString() != commentId).toList();
        print('✅ [CommentsNotifier] Commentaire supprimé avec succès');
        return true;
      }
      
      return false;
    } catch (error) {
      print('❌ [CommentsNotifier] Erreur suppression commentaire: $error');
      return false;
    }
  }

  void setContext(String contextType, String contextId) {
    // Cette méthode peut être utilisée pour stocker le contexte actuel
    print('📋 [CommentsNotifier] Contexte défini: $contextType:$contextId');
  }

  bool canDeleteComment(dynamic comment) {
    // Logique pour vérifier si l'utilisateur peut supprimer le commentaire
    // À implémenter selon votre logique d'authentification
    return false;
  }

  void clearCache() {
    state = [];
    print('🧹 [CommentsNotifier] Cache commentaires vidé');
  }
}

final useCommentsProvider = StateNotifierProvider<CommentsNotifier, List<dynamic>>((ref) {
  final apiService = ref.read(apiServiceProvider);
  return CommentsNotifier(apiService);
}); 